src/cargo/sources/path.rs

```diff
@@ -85,42 +87,161 @@ impl<'cfg> PathSource<'cfg> {
+    ///
+    /// ## Pattern matching strategy
+    ///
+    /// Migrating from a glob-like pattern matching (using `glob` crate) to a
+    /// gitignore-like pattern matching (using `ignore` crate). The migration
+    /// stages are:
+    ///
+    /// 1) Only warn users about the future change iff their matching rules are
+    ///    affected.  (CURRENT STAGE)
+    ///
+    /// 2) Switch to the new strategy and upate documents. Still keep warning
+    ///    affected users.
+    ///
+    /// 3) Drop the old strategy and no mor warnings.
+    ///
+    /// See <https://github.com/rust-lang/cargo/issues/4268> for more info.
     pub fn list_files(&self, pkg: &Package) -> CargoResult<Vec<PathBuf>> {
         let root = pkg.root();
+        let no_include_option = pkg.manifest().include().is_empty();
 
-        let parse = |p: &String| {
+        // glob-like matching rules
+
+        let glob_parse = |p: &String| {
             Pattern::new(p).map_err(|e| {
-                CargoError::from(format!("could not parse pattern `{}`: {}", p, e))
+                CargoError::from(format!("could not parse glob pattern `{}`: {}", p, e))
             })
         };
 
-        let exclude = pkg.manifest()
-                         .exclude()
-                         .iter()
-                         .map(|p| parse(p))
-                         .collect::<Result<Vec<_>, _>>()?;
-
-        let include = pkg.manifest()
-                         .include()
-                         .iter()
-                         .map(|p| parse(p))
-                         .collect::<Result<Vec<_>, _>>()?;
-
-        let mut filter = |p: &Path| {
-            let relative_path = util::without_prefix(p, root).unwrap();
-            include.iter().any(|p| p.matches_path(relative_path)) || {
-                include.is_empty() &&
-                 !exclude.iter().any(|p| p.matches_path(relative_path))
+        let glob_exclude = pkg.manifest()
+            .exclude()
+            .iter()
+            .map(|p| glob_parse(p))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let glob_include = pkg.manifest()
+            .include()
+            .iter()
+            .map(|p| glob_parse(p))
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let glob_should_package = |relative_path: &Path| -> bool {
+            // include and exclude options are mutually exclusive.
+            if no_include_option {
+                !glob_exclude.iter().any(|pattern| {
+                    pattern.matches_path(relative_path)
+                })
+            } else {
+                glob_include.iter().any(|pattern| {
+                    pattern.matches_path(relative_path)
+                })
+            }
+        };
+
+        // ignore-like matching rules
+
+        let mut exclude_builder = GitignoreBuilder::new(root);
+        for rule in pkg.manifest().exclude() {
+            exclude_builder.add_line(None, rule)?;
```

Comment:
Just to double check, how can this fail? This may be an error we wish to handle if valid `glob`-crate patterns are invalid gitignore patterns right now